### PR TITLE
[SPARK-21957][SQL] Support current_user function

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Currently, we do not have a suitable definition of the `user` concept in Spark. We only have a `sparkUser` app widely but do not support identify or retrieve the user information from a session in STS or a runtime query execution.

`current_user()` is very popular and supported by plenty of other modern or old school databases, and also ANSI compliant.

This PR add `current_user()`  as a SQL function. And, they are the same.  In this PR, we add these functions w/o ambiguity.
1. For a normal single-threaded Spark application, clearly the `sparkUser` is always equivalent to `current_user()` .
2. For a multi-threaded Spark application, e.g. Spark thrift server, we use a `ThreadLocal` variable to store the client-side user(after authenticated) before running the query and retrieve it in the parser.

### Why are the changes needed?

`current_user()` is very popular and supported by plenty of other modern or old school databases, and also ANSI compliant.

### Does this PR introduce _any_ user-facing change?

yes, added  `current_user()`  as a SQL function
### How was this patch tested?

new tests in thrift server and sql/catalyst

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @wangyum @maropu thanks very much

File: sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala [hunk 1]
```diff
@@ -40,6 +40,12 @@ class MiscFunctionsSuite extends QueryTest with SharedSparkSession {
       Row(SPARK_VERSION_SHORT + " " + SPARK_REVISION))
     assert(df.schema.fieldNames === Seq("version()"))
   }
+
+  test("get current_user and session_user in normal spark apps") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala:44]: shall we add the JIRA prefix?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/MiscFunctionsSuite.scala:44]: my bad, I forgot it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 2]
```diff
@@ -283,3 +283,21 @@ case class TypeOf(child: Expression) extends UnaryExpression {
 
   override protected def withNewChildInternal(newChild: Expression): TypeOf = copy(child = newChild)
 }
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """_FUNC_() - Return the current user who is running SparkContext for current query execution.""",
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:289]: As a SQL function, we need to be engine-agnostic and define the semantic without Spark.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:289]: right

**Comment by yaooqinn** [comment]
the commit 8c70c175455219096622653d37fe8642a722e9d1 blocks `SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/testOnly *ExpressionsSchemaSuite"` for updating golden file, will check this later
```
[error] Symbol 'type org.scalactic.TripleEquals' is missing from the classpath.
[error] This symbol is required by 'trait org.scalatest.Assertions'.
[error] Make sure that type TripleEquals is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
[error] A full rebuild may help if 'Assertions.class' was compiled against an incompatible version of org.scalactic.
[error] one error found
[info] compiling 27 Scala sources to /Users/kentyao/spark/sql/core/target/scala-2.12/classes ...
[error] (core / Test / compileIncremental) Compilation failed
[error] Total time: 40 s, completed 2021-6-1 15:58:05
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 3]
```diff
@@ -109,6 +112,11 @@ case class GetCurrentDatabaseAndCatalog(catalogManager: CatalogManager) extends
         Literal.create(currentNamespace, StringType)
       case CurrentCatalog() =>
         Literal.create(currentCatalog, StringType)
+      case CurrentUser() =>
+        val currentUser = SparkContext.getActive.map { sc =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:116]: nit:
```
SparkContext.getActive.flatMap { sc =>
  Option(sc.getLocalProperty(SESSION_USER_KEY))
}.getOrElse(Utils.getCurrentUserName())
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 4]
```diff
@@ -109,6 +112,11 @@ case class GetCurrentDatabaseAndCatalog(catalogManager: CatalogManager) extends
         Literal.create(currentNamespace, StringType)
       case CurrentCatalog() =>
         Literal.create(currentCatalog, StringType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:114]: like this, we should pre-calculate the current user before transforming the query plan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala [hunk 5]
```diff
@@ -98,17 +101,21 @@ object ComputeCurrentTime extends Rule[LogicalPlan] {
  * Replaces the expression of CurrentDatabase with the current database name.
  * Replaces the expression of CurrentCatalog with the current catalog name.
  */
-case class GetCurrentDatabaseAndCatalog(catalogManager: CatalogManager) extends Rule[LogicalPlan] {
+case class ReplaceCurrentLike(catalogManager: CatalogManager) extends Rule[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:104]: after this renaming, I think current_date etc. should also be calculated here. We can do it in a followup.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala:103]: ok, marked

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CurrentUserContext.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import org.apache.spark.util.Utils
+
+object CurrentUserContext {
+  val CURRENT_USER: InheritableThreadLocal[String] = new InheritableThreadLocal[String] {
+    override protected def initialValue(): String = Utils.getCurrentUserName()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CurrentUserContext.scala:24]: The default value can be null.

**Comment by yaooqinn** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**